Theming support for rich texts: rich text input and localized rich text input

ddouglasz opened the pull request:
Summary:

Theming support for rich texts: rich text input and localized rich text input

kark reviewed (commented):
I think it already looks good 🙌 just a few details I noticed - I'm not sure all of those are taken into account:
<img width="744" alt="Screenshot 2023-03-08 at 08 56 49" src="https://user-images.githubusercontent.com/49066275/223655914-aacff7b6-2d12-45df-a84b-5bb568896c93.png">

ddouglasz commented:
Updated here: https://github.com/commercetools/ui-kit/pull/2449/commits/a0494149c083968c53d284b2d2a5ee8082210cbc

Thanks!

emmenko commented:
Nit: I think the same should apply to readonly and disabled states (the text color).

ddouglasz commented:
updated here: https://github.com/commercetools/ui-kit/pull/2449/commits/adcc988aaca26762f026faa90d85f7953c86cc0e

Thanks!

kark reviewed (commented):
Thanks!
The generated `design-tokens.ts` file was not committed.
I suppose these are my closing remarks:
<img width="838" alt="Screenshot 2023-03-10 at 09 12 51" src="https://user-images.githubusercontent.com/49066275/224263276-c2cd4a2a-e1af-4ba9-8f4d-bf9e1fb92d6e.png">

ddouglasz commented:
> Thanks! The generated `design-tokens.ts` file was not committed. I suppose these are my closing remarks to the visual part: <img alt="Screenshot 2023-03-10 at 09 12 51" width="838" src="https://user-images.githubusercontent.com/49066275/224263276-c2cd4a2a-e1af-4ba9-8f4d-bf9e1fb92d6e.png">

Oops! thanks, updated: https://github.com/commercetools/ui-kit/pull/2449/commits/fb0dc2fcdf237b9b6f3e0fa50376f6fc3199321d

CarlosCortizasCT commented:
I think I spotted a couple of little things to adjust:
* Vertical spacing between actions and horizontal separator seems to be smaller than the one in the design
* The expand icon used in the designs is different than the one we're using right now

![image](https://user-images.githubusercontent.com/97907068/224324948-23f83b58-30ca-4374-ba1f-bcb72cf9ad3e.png)

ddouglasz commented:
> I think I spotted a couple of little things to adjust:
> 
> * Vertical spacing between actions and horizontal separator seems to be smaller than the one in the design
> * The expand icon used in the designs is different than the one we're using right now
> 
> ![image](https://user-images.githubusercontent.com/97907068/224324948-23f83b58-30ca-4374-ba1f-bcb72cf9ad3e.png)

Oh I guess something weird is happening that I need to check out.  Just noticed the vertical spacing acts like page needs to reload before showing. looking into it.
<img width="343" alt="image" src="https://user-images.githubusercontent.com/19975842/224326245-f1f179c8-50c5-447f-9b8a-dc294ebc9e0f.png">

ddouglasz commented:
Updated: https://github.com/commercetools/ui-kit/pull/2449/commits/33837ae08ba10654f7a8ce17ba3ef3eb941af6ed

Thanks!

CarlosCortizasCT commented:
> > I think I spotted a couple of little things to adjust:
> > 
> > * Vertical spacing between actions and horizontal separator seems to be smaller than the one in the design
> > * The expand icon used in the designs is different than the one we're using right now
> > 
> > ![image](https://user-images.githubusercontent.com/97907068/224324948-23f83b58-30ca-4374-ba1f-bcb72cf9ad3e.png)
> 
> Oh I guess something weird is happening that I need to check out. Just noticed the vertical spacing acts like page needs to reload before showing. looking into it. <img alt="image" width="343" src="https://user-images.githubusercontent.com/19975842/224326245-f1f179c8-50c5-447f-9b8a-dc294ebc9e0f.png">

It seems that padding is controlled by the `padding-for-rich-text-toolbar` token which currently has this values:
* default theme: `var(--spacing-10) 7px`
* new theme: `none`

It seems we would need to update the padding for the new theme, although I cannot find the right value when inspecting the Figma design. Perhaps @FilPob can help here.

ddouglasz commented:
> ... It seems we would need to update the padding for the new theme, although I cannot find the right value when inspecting the Figma design. Perhaps @FilPob can help here.

Oh yes I get you now. You mean we might need to give the new padding a value  asides the "none" as indicated on the design. Yeap, @FilPob can help with this :)

FilPob commented:
Yeah the toolbar should have padding-bottom: 4px and padding-top: 8px. This seems to be correct in the preview link

- Additionally I see that the height of the toolbar is too small. Each button should have a padding of 8px instead of 4px. 
- The Style dropdown should then have the same height as the buttons. Right now it has a different size 
<img width="133" alt="image" src="https://user-images.githubusercontent.com/94548534/224734259-af5ca2b4-08b9-4062-b335-9bea0162c825.png">

ddouglasz commented:
> Yeah the toolbar should have padding-bottom: 4px and padding-top: 8px. This seems to be correct in the preview link
> 
> * Additionally I see that the height of the toolbar is too small. Each button should have a padding of 8px instead of 4px.
> * The Style dropdown should then have the same height as the buttons. Right now it has a different size
> 
> <img alt="image" width="133" src="https://user-images.githubusercontent.com/94548534/224734259-af5ca2b4-08b9-4062-b335-9bea0162c825.png">

Thanks @FilPob , this has been taken care.
https://github.com/commercetools/ui-kit/pull/2449/commits/8c837ddb823988af8cbe858a56cfcd690e3ebabf

CarlosCortizasCT commented on packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.styles.tsx:
```diff
@@ -41,11 +41,11 @@ const getButtonStyles = (props: TDropdownStylesProps) => [
     padding: ${props.isStyleButton
-      ? `calc(${designTokens.spacing10} - 1px) ${designTokens.spacing20}`
-      : designTokens.spacing10};
+      ? `${designTokens.paddingForLocalizedRichTextDropdownButton} ${designTokens.spacing20}`
```
Could we include the `spacing20` within the new token?
It seems the the new token only includes vertical padding and I guess it should include both.

CarlosCortizasCT commented on packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.styles.tsx:
```diff
@@ -41,11 +41,11 @@ const getButtonStyles = (props: TDropdownStylesProps) => [
 
     &:hover {
-      background-color: ${designTokens.colorNeutral90};
+      background-color: ${designTokens.colorNeutral95};
```
It seems this change affects both themes. Is it on purpose?

CarlosCortizasCT commented on packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.styles.tsx:
```diff
@@ -55,7 +55,7 @@ const getButtonStyles = (props: TDropdownStylesProps) => [
     css`
       &:not(:hover) {
-        background-color: ${designTokens.colorAccent30};
+        background-color: ${designTokens.colorAccent20};
```
It seems this change affects both themes. Is it on purpose?

CarlosCortizasCT commented on packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body-button.tsx:
```diff
@@ -37,13 +37,14 @@ const RichTextBodyButton = (props: TRichTextBodyButtonProps) => {
           cursor: pointer;
           background: ${props.isActive
-            ? designTokens.colorAccent30
+            ? designTokens.colorAccent20
```
It seems this change affects both themes. Is it on purpose?

CarlosCortizasCT commented on packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body-button.tsx:
```diff
@@ -52,8 +53,8 @@ const RichTextBodyButton = (props: TRichTextBodyButtonProps) => {
             background: ${props.isActive
-              ? designTokens.colorAccent30
-              : designTokens.colorNeutral90};
+              ? designTokens.colorAccent20
+              : designTokens.colorNeutral95};
```
It seems this change affects both themes. Is it on purpose?

CarlosCortizasCT commented on packages/components/inputs/rich-text-utils/src/rich-text-body/rich-text-body.tsx:
```diff
@@ -341,7 +346,11 @@ const RichTextEditorBody = forwardRef<
-          <Divider />
+          <Divider
+            css={css`
+              ${isNewTheme ? 'margin: 8px 0' : 'inherit'}
```
Could we have a token for this use case?

CarlosCortizasCT commented:
We might want to increase the with of the `LocalizedRichTextInput` component is its story as it is too narrow in the new theme.

![image](https://user-images.githubusercontent.com/97907068/224962867-89c1e114-64a7-4239-8392-613c8663ac8d.png)

FilPob commented:
Thanks, looking better. Some more points I found looking into it more detailed:

1. The language label should have a padding left&right of 12px, same as other localized inputs

2. the 4px gap between buttons should be also applied around the divider. Meaning `more styles button` - 4px - `divider` - 4px -` bullet list button`
![image](https://user-images.githubusercontent.com/94548534/224977111-5e1a0cf9-7772-4e7c-a062-6f607a9bf3f1.png)

3. If you turn on the showExandIcon prop, the buttons are not aligned properly
![image](https://user-images.githubusercontent.com/94548534/224980126-3b3c589e-2dfa-4e61-ab5d-386766f3b750.png)

4. I would like to change the padding-bottom of the toolbar to 8px. Sorry for the back and forth, just realised it looks too squeezed with 4px
![image](https://user-images.githubusercontent.com/94548534/224979937-5511dfa1-ee9e-450a-a339-11b24ccbe348.png)

5. The dropdown menu for `More styles` should consume the same styles as found in the select inputs. I added the design **[HERE](https://www.figma.com/file/UoqmtHSHwxvEZRcbsM4A3Z/MC_Redesign_library?node-id=1387%3A24980&t=EIbjYHM3B1XmMNIa-1).** 
![image](https://user-images.githubusercontent.com/94548534/224985017-365fe720-8f0e-404b-93ce-2abdaa06db73.png)

ddouglasz commented:
Thanks @FilPob and @CarlosCortizasCT .
Feedback handled here: https://github.com/commercetools/ui-kit/pull/2449/commits/e4f63b01ea60b3ad3b9b0589f3bbba5ecf3e9cd4

CarlosCortizasCT commented:
> Thanks @FilPob and @CarlosCortizasCT . Feedback handled here: [e4f63b0](https://github.com/commercetools/ui-kit/commit/e4f63b01ea60b3ad3b9b0589f3bbba5ecf3e9cd4)

Thanks for the iterations @ddouglasz 👍 

After reviewing the UI again, I think I'm still missing Filip's point number 5.

Also, perhaps we can make the `LocalizedRichTextInput` a little bigger since all controls do not fit in one line when showing the expand icon in the new theme:

![image](https://user-images.githubusercontent.com/97907068/225078026-0d8c17e1-ac8d-4b68-a557-254078e09cca.png)

ddouglasz commented:
> ... After reviewing the UI again, I think I'm still missing Filip's point number 5.
> 
> Also, perhaps we can make the `LocalizedRichTextInput` a little bigger since all controls do not fit in one line when showing the expand icon in the new theme:

Ah, thanks : https://github.com/commercetools/ui-kit/pull/2449/commits/f8adde0e7cfdd9821f0f2a279a20ed97eda71735

CarlosCortizasCT commented:
Regarding the "More styles" dropdown, I think we still need to adjust the spacings and the spacings, border and shadow.

This is how it looks now:
![image](https://user-images.githubusercontent.com/97907068/225241712-9881fe5c-5935-4ca4-98d2-0bb15150a79f.png)

This is how it looks in the [Figma design](https://www.figma.com/file/UoqmtHSHwxvEZRcbsM4A3Z/MC_Redesign_library?node-id=1387%3A24980&t=ounXioveGpoTkpNn-0):
![image](https://user-images.githubusercontent.com/97907068/225242148-aa8b2e0e-64bf-4d1c-a1d6-8d1e194bc2c5.png)

Also, there is an _active_ state in Figma that we-re not currently managing in the dropdown.

kark reviewed (commented):
This is really great work 👏 Two other minor issues I found:
- seems there are no specific `hover` state styles
<img width="608" alt="Screenshot 2023-03-15 at 09 05 06" src="https://user-images.githubusercontent.com/49066275/225246861-0b19db06-5e7e-41e8-ab94-7595f50f084b.png">

- The language controls changes color to grey when in `hasError` or `hasWarning` is truthy
<img width="628" alt="Screenshot 2023-03-15 at 09 08 15" src="https://user-images.githubusercontent.com/49066275/225247552-0a1913ff-fa61-47a7-aa4f-2186216f8f9b.png">

ddouglasz commented:
> This is really great work 👏 Two other minor issues I found:
> 
> * seems there are no `hover` state specific styles
> 
> <img alt="Screenshot 2023-03-15 at 09 05 06" width="608" src="https://user-images.githubusercontent.com/49066275/225246861-0b19db06-5e7e-41e8-ab94-7595f50f084b.png">
> 
> * The language control changes color to grey when in `hasError` or `hasWarning` is truthy
> 
> <img alt="Screenshot 2023-03-15 at 09 08 15" width="628" src="https://user-images.githubusercontent.com/49066275/225247552-0a1913ff-fa61-47a7-aa4f-2186216f8f9b.png">

Thanks @kark, really awesome observations, once again 🙏🏾

1. To the 1st point, I think we may have to consult with @FilPob on that. My reason being if `on-hover`, we want the entire field(Including the editor top part to have a background color change, or just where the input field exists.

2. To the 2nd point, I see its the same behaviour everywhere else, `error` or `warning` state greys out the language toggle.

ddouglasz commented:
> Regarding the "More styles" dropdown, I think we still need to adjust the spacings and the spacings, border and shadow.
> 
> ...
> Also, there is an _active_ state in Figma that we-re not currently managing in the dropdown.

updated here: thank you. I also added some slight figma changes from Filip. basically covering the style dropdown too: https://github.com/commercetools/ui-kit/pull/2449/commits/3c85647325213f8c728a68199479acbf3eb62f31

kark commented:
> 2. To the 2nd point, I see its the same behaviour everywhere else, `error` or `warning` state greys out the language toggle.

Oh, that's right, thanks 👍

ddouglasz commented:
@kark to your first point, I have consulted with FIlip and the conclusion is to add the `on-hover` for the entire component. Apparently, this had already been implemented but had a tiny bug, which I fixed here: https://github.com/commercetools/ui-kit/pull/2449/commits/ba0d49233866c3e527f8795fc372cd15e50b9aa9

CarlosCortizasCT commented on packages/components/inputs/rich-text-utils/src/rich-text-body/dropdown.tsx:
```diff
@@ -108,7 +109,13 @@ const Dropdown = (props: TDropdownProps) => {
                         isSelected={props.getIsItemSelected(item)}
                       >
-                        {item.label}
+                        <div
```
**nit** perhaps we can have a private styled component for this instead of using inline CSS:
```
const DropdownItemLabelWrapper = styled.div`
  margin: ${designTokens.marginForRichTextDropdownItemLabel};
`;
```

ddouglasz replied:
> AWESOME 💪
> 
> One last thing from my end: there is a tiny issue with `cursor: not-allowed` ![2023-03-15 16 41 50](https://user-images.githubusercontent.com/49066275/225363884-92fc10af-2b43-44d3-8702-7b747036ecf9.gif) [ ![2023-03-15 16 41 50](https://user-images.githubusercontent.com/49066275/225363884-92fc10af-2b43-44d3-8702-7b747036ecf9.gif) ](https://user-images.githubusercontent.com/49066275/225363884-92fc10af-2b43-44d3-8702-7b747036ecf9.gif) [ ](https://user-images.githubusercontent.com/49066275/225363884-92fc10af-2b43-44d3-8702-7b747036ecf9.gif)

Yes, this is also happening on the default behaviour theme if you zoom in and hover over the space between two editor buttons. I can do a follow up to see why this is happening.

FilPob commented:
Awesome! There is one more last thing I see and that is that the toolbar still has a doubled padding on the top. Should be 8px Otherwise everything perfect! 
![image](https://user-images.githubusercontent.com/94548534/225355872-056aaa32-b8d9-4f2a-a95e-3bb872a0c3ab.png)

kark reviewed (approved):
AWESOME 💪

One last thing from my end: there is a tiny issue with `cursor: not-allowed`
![2023-03-15 16 41 50](https://user-images.githubusercontent.com/49066275/225363884-92fc10af-2b43-44d3-8702-7b747036ecf9.gif)

ddouglasz commented:
feedback updated here: https://github.com/commercetools/ui-kit/pull/2449/commits/a2402bce1cb9a250e1346abe926dfc0af68cff40 and https://github.com/commercetools/ui-kit/pull/2449/commits/8a923af5c6094c231c2351a9f15ddf895b25c536

FilPob commented:
@ddouglasz  awesome! looks perfect now !

chloe0592 reviewed (approved):
👏 👏
